src/cuttlefish_generator.erl

```diff
@@ -45,9 +45,48 @@ map(Schema, Config) ->
 
 map(Schema, Config, ConfFile) ->
     IncludeFile = lists:filter(fun({K, _}) -> K =:= include_file end, Config),
-    AllConfig = merge_include_conf(Config -- IncludeFile, IncludeFile, ConfFile),
+    IncludedConfig = merge_include_conf(Config -- IncludeFile, IncludeFile, ConfFile),
+    AllConfig = merge_env_conf(IncludedConfig, Schema),
     map_add_defaults(Schema, AllConfig).
 
+-spec (merge_env_conf(cuttlefish_conf:conf(), cuttlefish_schema:schema()) -> cuttlefish_conf:conf()).
+merge_env_conf(Config, Schema) ->
+    Envs = match_schema(getenv(), Schema),
+    lists:ukeymerge(1, lists:sort(Envs), lists:sort(Config)).
+
+-spec (getenv() -> cuttlefish_conf:conf()).
+getenv() ->
+    Envs = lists:map(fun (E) -> string:split(E, "=", leading) end, os:getenv()),
+    do_getenv(Envs, []).
+
+do_getenv([], Acc) ->
+    lists:sort(Acc);
+do_getenv([["EMQX_" ++ Key, Value]|More], Acc) ->
+    FormattedKey = string:lowercase(string:replace(Key, "__", ".", all)),
+    do_getenv(More, [{cuttlefish_variable:tokenize(FormattedKey), Value}|Acc]);
+do_getenv([_Other|More], Acc) ->
+    do_getenv(More, Acc).
+
+-spec(match_schema(cuttlefish_conf:conf(), cuttlefish_schema:schema()) -> cuttlefish_conf:conf()).
+match_schema(Envs, {_, Mappings, _}) ->
+    ValidKeys = lists:map(fun (Mapping) -> {element(2, Mapping), '_'} end, Mappings),
+    do_match_schema(lists:sort(Envs), lists:sort(ValidKeys), []).
```

Comment:
maybe more efficient if we make a `set()` from `lists:map(fun(Mapping) -> element(2, Mapping) end, Mappings)` instead of `lists:sort`.

and there is no need to sort `Envs` either.